third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,248 @@
+    // factorizations of numWarps. Pick shape for which both conversions in
+    // decomposition use LDS less than LDSSize and for which sum of LDS usage
+    // is minimal. If no such shape exists, do not decompose.
+    auto factorizedNumWarps =
+        mlir::triton::AMD::factorizePowerOf2(numWarps, rank);
+    // Create a list of temporary layouts
+    SmallVector<Attribute> tmpLayouts;
+    for (int i = 0; i < factorizedNumWarps.size(); i++) {
+      auto warpsPerCTA = factorizedNumWarps[i];
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(srcEnc, warpsPerCTA));
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(dstEnc, warpsPerCTA));
+      SmallVector<unsigned> elemsPerThread(rank, 1);
+      SmallVector<unsigned> threadsPerWarp(rank, 1);
+      threadsPerWarp[rank - 1] = warpSize / 8;
+      threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+      auto order = triton::gpu::getOrder(srcEnc);
+      auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+      auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
+          ctx, elemsPerThread, threadsPerWarp, warpsPerCTA, order, layoutCTA);
+      tmpLayouts.push_back(fallbackLayout);
+    }
+
+    unsigned minLDSUsage = 2 * LDSSize;
+    int minIdx = -1;
+    for (int i = 0; i < tmpLayouts.size(); i++) {
+      auto resources = mlir::triton::AMD::estimateResourcesForReplacement(
+          builder, cvtOp, tmpLayouts[i]);
+      // TODO analyze performance along with LDS consumption
+      if (resources.LDS < minLDSUsage) {
+        minLDSUsage = resources.LDS;
+        minIdx = i;
+      }
+    }
+
+    if (minIdx == -1 || minLDSUsage > targetLDSSize) {
+      return;
+    }
+
+    triton::gpu::ConvertLayoutOp tmpCvt;
+    triton::gpu::ConvertLayoutOp newEpilogueCvt;
+
+    assert(minIdx >= 0 && minIdx < tmpLayouts.size());
+    auto tmpLayout = tmpLayouts[minIdx];
+    std::tie(tmpCvt, newEpilogueCvt) =
+        mlir::triton::AMD::createNewConvertOps(builder, cvtOp, tmpLayout);
+
+    cvtOp.replaceAllUsesWith(newEpilogueCvt.getResult());
+    cvtOp.erase();
+  }
+
+  struct LDSBottleneckOperation {
+    triton::gpu::ConvertLayoutOp op;
+    int64_t LDSSizeTarget;
+  };
+
+  /**
+   * Assuming that all buffer above scratch buffer in memory space can be
+   * shifted down in memory, this function gives optimistic estimation of memory
+   * space available for scratch buffer.
+   */
+  int64_t
+  computeMaxScratchBufferSize(triton::gpu::ConvertLayoutOp op,
+                              Allocation *allocation,
+                              ArrayRef<Allocation::BufferId> liveBuffers) {
+    int totalSize = 0;
+    auto scratchBufferId = allocation->getBufferId(op.getOperation());
+    int64_t scratchBufferSize = allocation->getAllocatedSize(scratchBufferId);
+    size_t totalLDSConsumption = 0;
+    for (auto buf : liveBuffers)
+      totalLDSConsumption = std::max(
```

Comment:
If all liveBuffers are live at this cvtOp, should we use sum instead of max here?

Reply:
Max is more conservative metric in this sense. Let's consider that we have "holes" in memory:

![Screenshot from 2024-05-08 16-12-09](https://github.com/openai/triton/assets/3631351/90a1f845-0d0a-48cc-967d-fe95f557e2bc)

let's consider that green buffer is scratch buffer that we want to optimize, viollet and blue are long-living buffers in shared layout.

Hole is created, because pink tensor is allocated on tick 1 and reallocated on tick 2, but previously allocated violet tensor continue live.

Summarizing buffer sizes will tell that we have 20 KB(3 * 8 KB) for scratch buffer, but in reality we probably wan to make it smaller.